### PR TITLE
Online authentication integration

### LINKED ISSUE #14: Online auth system testing
**Describe the bug**
The online mode bypass doesn't work.

**To Reproduce**
1. Create a server in offline mode with the mod installed, and sessions disabled.
2. Join the server with a Minecraft account, register, and leave the server
3. Join the server again with the same account.
4. Mod still asks to authenticate by the normal way.

**Expected behavior**
The mod should authenticate the player when it has already registered and has a Minecraft account.

**Version**
Mod version: 1.2.0 (I'm using the mod built from #13)
Minecraft version: 1.16.5
Fabric version: 0.36.0+1.16

**The reason**
<details>
<summary>Click to expand</summary>

(I have studied this part of Minecraft before when I tried to create an auth mod with online mode authentication.)
The reason this happens is simple: when the server is in offline mode, **all** players are assigned an UUID based on their username (see `PlayerEntity#getOfflinePlayerUuid`). (The UUID for all players is generated by the server, not by the client). Your code doesn't work because it compares the uuid generated by the **offline** server with the official one, which are always different.
</details>

**The solution**
<details>
<summary>Click to expand</summary>

I can only think of three solutions.
- Reimplementing Mojang's/Minecraft's authentication to check if the player has a proper account. This would be overly complicated and it may not even be possible.
- When the player logs on to the server (`ServerLoginNetworkHandler#onHello`), checking if a Minecraft account with the same username exists. If so, making the Minecraft server handle the authentication and if not, just letting the player pass and authenticate the normal way (`/login`). This would cause players with the username of a valid Minecraft account to not be able to join the server.
- Letting each player opt-in to the online mode bypass feature. Like the previous solution, when the player logs on to the server, checking if the player has opted in to the feature. If the player has opted into the feature, let Minecraft handle the authentication part and, if not, let the player authenticate the normal way.
Do note that the second and third would make players using an official account to use their online-mode-uuid, so they would be able to change their account username and save all their data, but losing the data from when they didn't have the option toggled on.

The letting pass an offline player or authenticate an online one part can be done with just a single mixin
<details>
<summary>Code</summary>

```java
@Mixin(ServerLoginNetworkHandler.class)
public class ServerLoginNetworkHandlerMixin {
    @Redirect(method = "onHello", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/MinecraftServer;isOnlineMode()Z"))
    private boolean authenticatePlayerWithMojang(MinecraftServer server) {
        if (!server.isOnlineMode() || !featureEnabled) return false;

        boolean playerHasFeatureEnabled = ...;
        return playerHasFeatureEnabled && playerHasMinecraftAccount(this.profile.getName());
    }

    private static boolean playerHasMinecraftAccount(String username) {
        // Code from OnPlayerConnect#testPlayerOnline, just checking if the server does return an uuid
    }
}
```
</details>
</details>

### PR DESCRIPTION
**Pull request by IotaBread**
Removes the "skip online auth." feature and introduces two replacements. Closes #14

(I'm going to refer to players with a valid license & account for Minecraft as 'official' and as 'unofficial' to those that don't have a license or account.)

- Forced online auth. - Forces all players using an username of an official player to use a Minecraft account. Prevents unofficial players from "stealing" the name of an official player (if the latter hasn't registered in the server).
  - EDIT: Added a config option `forced-online-users` (an array of usernames) to select which players should use offline auth even if there is an official player with that name.
- Optional online auth. - Lets official players to choose to use online mode authentication with the `/onlineauth` command, effectively skipping `/login` authentication.

**Note**: Both features require the server to use online mode, although I think it could work in offline mode (if the `if` statements checking the online mode are removed from the code).
**Note**: Both features give official players their UUID given by Mojang, allowing them to change their account username without losing data (in optional online auth. they have to run `/onlineauth` again). Unofficial players keep their offline-mode UUID.

I don't know Russian, so I left the added messages in English

### CODE REVIEW

**Comment by IotaBread** [comment]
Apparently it doesn't work outside dev. Gonna fix it tomorrow

File: src/main/java/com/dqu/simplerauth/AuthMod.java [hunk 1]
```diff
@@ -1,28 +1,74 @@
...
+                }
+                content = stringBuilder.toString();
+            } else return false;
+        } catch (Exception e) {
+            AuthMod.LOGGER.error(e);
+            return false;
+        }
+
+        JsonObject jsonObject = GSON.fromJson(content, JsonObject.class);
+        if (jsonObject.has("id")) {
```
Comment (Dqu1J) [inline src/main/java/com/dqu/simplerauth/AuthMod.java]: Not needed json object, as the Mojang API will return code 204 if account doesn't exist.

File: src/main/java/com/dqu/simplerauth/managers/CacheManager.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            String json = GSON.toJson(db);
+            bufferedWriter.write(json);
+            bufferedWriter.close();
+        } catch (Exception e) {
+            AuthMod.LOGGER.error(e);
+        }
+    }
+
+    public static JsonArray getMinecraftAccounts() {
+        return db.getAsJsonArray("minecraft-account-cache");
```
Comment (Dqu1J) [inline src/main/java/com/dqu/simplerauth/managers/CacheManager.java]: Should be `db.get("minecraft-account-cache").getAsJsonArray()`

File: src/main/java/com/dqu/simplerauth/managers/CacheManager.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    public static void addMinecraftAccount(String username, String onlineUuid) {
+        String offlineUuid = PlayerEntity.getOfflinePlayerUuid(username).toString();
+        onlineUuid = formatUuid(onlineUuid);
+
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty("username", username);
+        jsonObject.addProperty("online-uuid", onlineUuid);
+        jsonObject.addProperty("offline-uuid", offlineUuid);
```
Comment (Dqu1J) [inline src/main/java/com/dqu/simplerauth/managers/CacheManager.java]: Isn't needed, can get that by `.getOfflinePlayerUuid()` to use less storage.

File: src/main/java/com/dqu/simplerauth/managers/CacheManager.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            JsonObject account = minecraftAccounts.get(i).getAsJsonObject();
+            LocalDateTime parsedTimestamp;
+            try {
+                parsedTimestamp = LocalDateTime.parse(account.get("timestamp").getAsString());
+            } catch (Exception e) {
+                continue;
+            }
+            boolean valid = Duration.between(parsedTimestamp, LocalDateTime.now()).toDays() <= 14;
+            if (!valid) {
+                invalidAccounts.add(account);
```
Comment (Dqu1J) [inline src/main/java/com/dqu/simplerauth/managers/CacheManager.java:96]: You can just remove it here?

File: src/main/java/com/dqu/simplerauth/managers/CacheManager.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        saveCache();
+    }
+
+    private static String formatUuid(String uuid) {
+        // Check if UUID doesn't have dashes
+        Pattern pattern = Pattern.compile("^[a-z0-9]{32}$");
+        Matcher matcher = pattern.matcher(uuid);
+        if (!matcher.matches()) return uuid;
+
+        return String.format("%s-%s-%s-%s-%s", uuid.substring(0, 8), uuid.substring(8, 12), uuid.substring(12, 16), uuid.substring(16, 20), uuid.substring(20));
```
Comment (Dqu1J) [inline src/main/java/com/dqu/simplerauth/managers/CacheManager.java]: `uuid.replace('-', '');`

File: src/main/java/com/dqu/simplerauth/managers/ConfigManager.java [hunk 6]
```diff
@@ -25,9 +27,10 @@ public static void loadConfig() {
 
             db.addProperty("sessions-enabled", true);
             db.addProperty("sessions-valid-hours", "6");
-            db.addProperty("skip-online-auth", true);
             db.addProperty("password-type", "local");
             db.addProperty("global-password", "123456");
+            db.addProperty("forced-online-auth", false);
```
Comment (Dqu1J) [inline src/main/java/com/dqu/simplerauth/managers/ConfigManager.java:32]: Forced online auth doesn't make sense, as there is online mode.

File: src/main/java/com/dqu/simplerauth/managers/ConfigManager.java [hunk 7]
```diff
@@ -4,13 +4,15 @@
...
 import net.fabricmc.loader.api.FabricLoader;
 
 import java.io.BufferedReader;
 import java.io.BufferedWriter;
 import java.io.File;
 import java.nio.charset.StandardCharsets;
+import java.util.Locale;
 
 public class ConfigManager {
     public static final int VERSION = 1;
```
Comment (Dqu1J) [inline src/main/java/com/dqu/simplerauth/managers/ConfigManager.java]: Should be version 2, and update the config to the new format if it is version 1.

File: src/main/java/com/dqu/simplerauth/managers/ConfigManager.java [hunk 8]
```diff
@@ -108,4 +111,15 @@ public static String getAuthType() {
         return authtype.equalsIgnoreCase("global") ? "global" : "none";
     }
 
+
+    public static boolean forcePlayerOffline(String username) {
+        if (!db.has("forced-offline-users")) return false;
```
Comment (Dqu1J) [inline src/main/java/com/dqu/simplerauth/managers/ConfigManager.java]: `forced-offline-users` isn't added to the db anywhere.

**Review by Dqu1J — CHANGES_REQUESTED** [review]

File: src/main/java/com/dqu/simplerauth/managers/CacheManager.java (hunk 4 above)
Comment (IotaBread) [reply src/main/java/com/dqu/simplerauth/managers/CacheManager.java:96]: Wouldn't that cause a CME?

Comment (Dqu1J) [reply src/main/java/com/dqu/simplerauth/managers/CacheManager.java:104]: > Wouldn't that cause a CME?

I'm sorry, what does CME stand for?

Comment (IotaBread) [reply src/main/java/com/dqu/simplerauth/managers/CacheManager.java:96]: ConcurrentModificationException. Usually happens when modifying something from an array that is being iterated over, but I'm not sure if it does happen with JsonArrays

Comment (Dqu1J) [reply src/main/java/com/dqu/simplerauth/managers/CacheManager.java:96]: It may. My bad.

**Comment by ghost** [comment]
I've tested all possible situations that can happen like stolen name conflict, original name conflict, attaching original name after registering non-original name etc. and everything works.. great mod improvement

**Review by Dqu1J — APPROVED** [review]
Will test it later, thanks for contributing!

**Comment by Dqu1J** [comment]
Please, see my message #14

**Pull request merged**